cogfiles/reaction_roles.py

```diff
@@ -0,0 +1,139 @@
+    message: discord.Message = field(compare=False)
+    role: discord.Role = field()
+    emoji: discord.PartialEmoji = field(compare=False)
+
+    def to_json(self) -> dict[str, int | str]:
+        """Convert the current ReactionRole object to a JSON string."""
+        return {
+            "message_id": self.message.id,
+            "channel_id": self.channel.id,
+            "role_id": self.role.id,
+            "emoji": str(self.emoji)
+        }
+
+    @staticmethod
+    async def from_json(bot: commands.Bot, json_obj: dict[str, int | str]):
+        """Convert a JSON dictionary to a ReactionRole object."""
+        channel = await bot.fetch_channel(json_obj["channel_id"])
+        
+        if channel is not None:
+            message = await channel.fetch_message(json_obj["message_id"])
+            role = channel.guild.get_role(json_obj["role_id"])
+            emoji = discord.PartialEmoji.from_str(json_obj["emoji"])
+            return ReactionRole(channel, message, role, emoji)
+
+class ReactionRoles(commands.Cog, name="Reaction Roles"):
+    """Manage ping roles through message reactions."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+        self.reactionroles: dict[int, set[ReactionRole]] = {}
+
+    @commands.Cog.listener()
+    async def on_ready(self):
+        """Initialize the list of reaction roles in memory."""
+        await initialize_from_json(self.bot, self.reactionroles, ReactionRole, "reaction_roles")
+    
+    @app_commands.command()
+    @app_commands.rename(emoji_str="emoji")
+    @app_commands.checks.has_permissions(manage_guild=True)
+    async def reactionrole(self, interaction: discord.Interaction, role: discord.Role, emoji_str: str, message_link: str):
+        "Attaches a reaction role to a message. When a user reacts to the message with the given emoji, they will receive the role."
+        emoji = discord.PartialEmoji.from_str(emoji_str)
+
+        regex_searches = re.search("(\d+)\/(\d+)$", message_link)
+        try:
+            channel_id = regex_searches.group(1)
+            message_id = regex_searches.group(2)
+            channel = await interaction.guild.fetch_channel(int(channel_id))
+            message = await channel.fetch_message(int(message_id))
+        except:
+            return await interaction.response.send_message("Could not find the given message.", ephemeral=True)
+
+        try:
+            await message.add_reaction(emoji)
+        except discord.errors.HTTPException as e:
+            return await interaction.response.send_message("Could not use the given emoji. Make sure that this bot shares a server with the emoji.", ephemeral=True)
+
+        reactionrole = ReactionRole(channel, message, role, emoji)
+
+        # If there already exists a reaction role for this role, remove it and replace it with the new one
+        old_reactionrole = next((rr for rr in self.reactionroles[interaction.guild.id] if rr.role == reactionrole.role), None)
```

Comment:
There's a couple places where we're using the abbreviation `rr` while iterating through the bot's list of reaction roles, which is a bit confusing when compared to a different variable named `reactionrole`. Perhaps we could rename `rr` to `role` and `reactionrole` to `new_reactionrole` in this case? (this would also line up with the name `old_reactionrole`)